shared-libs/cht-datasource/src/qualifier.ts

```
@@ -140,3 +140,66 @@ export const and = <
     ...(qualifierD ?? {}),
   };
 };
+
+/** 
+ * A qualifier for a contact
+ */
+type ContactQualifier = Readonly<{
+  type: string,
+  name: string,
+  reported_date?: string | number,
+  _id?: string,
+  _rev?: string
+}>;
+
+/**
+ * Builds a qualifier for creation and updation of a contact with
+ * the given fields.
+ * @param data object containing the fields for a contact
+ * @returns the contact qualifier
+ * @throws Error if type is not provided or is empty
+ * @throws Error if name is not provided or is empty
+ * @throws Error if reported_date is not in a valid format. 
+ * Valid formats are 'YYYY-MM-DDTHH:mm:ssZ', 'YYYY-MM-DDTHH:mm:ss.SSSZ', or <unix epoch>.
+ */
+export const byContactQualifier = (data: object) :ContactQualifier => {
+  const qualifier = {...data};
+  if (!isContactQualifier(qualifier)){
+    throw new InvalidArgumentError(`Missing required fields [${JSON.stringify(data)}]`);
+  }
+  return qualifier;
+};
+
+/**
+ * Returns `true` if the given qualifier is a {@link ContactQualifier} otherwise `false`.
+ * @param qualifier the qualifier to check
+ * @returns `true` if the given type is a {@link ContactQualifier}, otherwise `false`.
+ */
+const isContactQualifier = (qualifier: unknown) : qualifier is ContactQualifier => {
+  if (isRecord(qualifier) && hasFields(qualifier, [{name: 'type', type: 'string'}, {name: 'name', type: 'string'}])){
+    if ('reported_date' in qualifier){
+      const date = qualifier.reported_date;
+      if (!isValidReportedDate(date)) {
+        throw new InvalidArgumentError(
+          `Invalid reported_date format. Expected 'YYYY-MM-DDTHH:mm:ssZ', 'YYYY-MM-DDTHH:mm:ss.SSSZ', or a Unix epoch.`
+        );
+      }
+    }
```

Just returning `false` from `isContactQualifier` might not be helpful to determine what kind of error to throw in `byContactQualifer`. Which is why I throw an error here, which doesn't seem to be the best way to do this.
One solution might be doing the type checks and date formatting individual checks inside `byContactQualifer` without reusing `isContactQualifier`, to prevent ambiguity regarding which error to throw for a false return.

---

I believe we want to throw a different error based on the input.
I've identified two types:
1. Missing required `name` or `type` fields.
2. Invalid date format.

---

This was discussed on today's call and here is the summary:
1. We perform the date validation before calling `isContactQualifier` in `byContactQualifier` to throw the correct exception.
2. We will only return a boolean value from `isContactQualifier` similar to other `is.*` functions.

Thanks @sugat009 for unblocking me on this.